Fluid Place Block Event

MoxLotus opened the pull request:
Adds an event that allows a modder to change the output of a cobblestone generator. This event also triggers when water and lava are mixed to create stone or obsidian.

I created this event because while using IWorldEventListener::notifyBlockUpdate is workable, I'm essentially blind to whether or not the block placed in the world was actually placed by mixing liquids. This event is a much more robust solution.

akros2 committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into 1.12.x

akros2 committed:
A new event for cobblestone generators.

FluidMixEvent allows a modder to change the output of a cobblestone
generator. This event is fired whenever two liquids mix to create a new
block.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4627) <br/>All committers have signed the CLA.

Daomephsta commented on src/test/java/net/minecraftforge/debug/FluidMixTest.java:
```diff
@@ -0,0 +1,25 @@
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+@Mod(modid = "fluidmixtest", name = "FluidMixTest", version = "0.0.0")
```
Needs `acceptableRemoteVersions = "*"` in the `@Mod` annotation, otherwise it'll prevent vanilla clients connecting to Forge servers launched from a dev env for working on Forge, which makes testing vanilla-Forge compatiblity more difficult.

Daomephsta commented on src/test/java/net/minecraftforge/debug/FluidMixTest.java:
```diff
@@ -0,0 +1,25 @@
+@Mod(modid = "fluidmixtest", name = "FluidMixTest", version = "0.0.0")
+@Mod.EventBusSubscriber @SuppressWarnings("unused")
+public class FluidMixTest
+{
```
This test mod should be disableable and disabled by default, so that it doesn't interfere with testing of other test mods or creation of other PRs. The conventional way of doing that is to have a `private static final boolean` which defaults to false, and is checked before doing anything. To use the test mod, the source is edited so the boolean is true.

mezz commented on patches/minecraft/net/minecraft/block/BlockDynamicLiquid.java.patch:
```diff
@@ -9,15 +9,25 @@
+             if (this.field_149764_J == Material.field_151587_i && p_180650_1_.func_180495_p(p_180650_2_.func_177977_b()).func_185904_a() == Material.field_151586_h)
+             {
+-                p_180650_1_.func_175656_a(p_180650_2_.func_177977_b(), Blocks.field_150348_b.func_176223_P());
++                //worldIn.setBlockState(pos.down(), Blocks.STONE.getDefaultState());
```
Do not comment out lines in patch files, you can just remove it.

mezz commented on patches/minecraft/net/minecraft/block/BlockLiquid.java.patch:
```diff
@@ -1,6 +1,25 @@
+                 if (integer.intValue() == 0)
+                 {
+-                    p_176365_1_.func_175656_a(p_176365_2_, Blocks.field_150343_Z.func_176223_P());
++                    //worldIn.setBlockState(pos, Blocks.OBSIDIAN.getDefaultState());
```
Do not comment out lines in patch files, you can just remove it.

mezz commented on patches/minecraft/net/minecraft/block/BlockLiquid.java.patch:
```diff
@@ -1,6 +1,25 @@
+                 if (integer.intValue() <= 4)
+                 {
+-                    p_176365_1_.func_175656_a(p_176365_2_, Blocks.field_150347_e.func_176223_P());
++                    //worldIn.setBlockState(pos, Blocks.COBBLESTONE.getDefaultState());
```
Do not comment out lines in patch files, you can just remove it.

mezz commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,6 +283,44 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+        }
+
+        @Override
+        public IBlockState getState()
```
For consistency, this should probably be called `getBlockState`, the field can be `blockState`, and the setter can be `setBlockState`.

MoxLotus replied:
I used FluidMixEvent::getState in hopes of preventing the casual modder from accessing BlockEvent::getState which I have rendered irrelevant.

covers1624 replied:
What about `mixState` then?

MoxLotus replied:
Renamed it to `newState`, `getNewState` and `setNewState`.

mezz commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,6 +283,44 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+        private final Material liquid2;
+        private IBlockState block;
+
+        public FluidMixEvent(World world, BlockPos pos, Material liquid1, Material liquid2, IBlockState block)
```
This seems to only be fired for `WATER` and `LAVA`. Are there other cases where fluids mix? Can two fluids with material `WATER` ever mix and fire an event?

Also it might be better to pass the fluid block positions instead of materials so that someone can check the properties of their fluid and create more advanced mixes.

MoxLotus replied:
Not in vanilla.

Minecraft doesn't necessarily track the positions by default, and it is possible that lava could mix with water coming from two directions. Perhaps if I provided the position of the block that triggered the event alongside the position of the block to be placed. A mod author can use those if they need to check for more complicated interactions.

By that logic the materials of the liquids in question should be pretty easy to determine. I think removing those and adding a second BlockPos for the block that triggered the event is the way to go.

mezz reviewed (changes requested):
Looking pretty good.

akros2 committed:
Changed FluidMixEvent to FluidPlaceBlockEvent.

The event will now additionally trigger when fire is placed by lava. I
decided against firing the event when water source blocks are created by
water because this is already handled by CreateFluidSourceEvent.

Additionally, instead of providing the materials of the liquids
involved, the event now simply records the position of the liquid than
triggered the event.

Removed comments from .java.patch files.
Disabled the test mod. Can be re-enabled by setting ENABLED to true.

HenryLoenwind commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,7 +283,38 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
-    
+
+    /**
+     * Fired when a liquid places a block.
```
This needs documentation for which cases this fires from vanilla. While I would have expected obsidian, cobble and stone, the fire surprised me.

jabelar replied:
Also, it would be good to clarify in the comments how to "cancel" the event. Since it isn't actually cancelable and also doesn't use a Result return value, I guess if you want to prevent the placement then you return the actual state of the block that is already there, correct?

tomelfring commented:
This looks somewhat related to #3355

HenryLoenwind commented:
@thommy101, yes it does. However, this one actually provides the reasoning Lex was missing in the replace event. It may be a good idea to add a superclass to this event, with the fields a generic replace event would have. Then all the other replacings could be added like this one over time. (smaller cans, less worms)

MoxLotus commented:
@thommy101 @HenryLoenwind 

This event is much more targeted, focusing on just fluids. In my mind it is parallel to PlaceEvent (which could now be called PlayerPlaceEvent). If superclassing were to be done with these it would probably have to be done with a major update. Someone with more forge experience can tackle that if they really want to.

The creation of water source blocks is already governed by CreateFluidSourceEvent and placing "flowing" blocks is too fundamental for me to feel comfortable interfering.  Admittedly, I'm not sure if it's even justifiable to fire an event when lava places fire, but I went ahead because it's the only other time I can think of when a liquid places a block.

Point being I don't want to add further complexity. I'll make one more commit soon with better documentation and I'm hoping to be done.

akros2 committed:
Reviewed Documentation

Added a (hopefully) better description to FluidPlaceBlockEvent. Changed
the name of a getter/setter for clarity.

jabelar commented:
Is water flowing into concrete powder covered by this event as well?

MoxLotus commented:
@jabelar Nope.

Like `BlockFarmland` or `BlockReed`, `BlockConcretePowder` is responsible for checking conditions and making changes. This event only fires when the liquid is responsible for changing block states.

This distinction may be pedantic, but right now I'm inclined to stick with it. It gives the scope of this event a firm boundary.

jabelar commented:
I do kinda feel that leaving out concrete is in fact a bit pendantic. Sure it was coded the other way around, but that is pretty arbitrary. But someone can always add it later so I suppose no big deal.

akros2 committed:
Made event cancelable.

Just in case someone wants to cancel/uncancel the event.

Also fixed a bug in FluidPlaceBlockTest created by the last commit.

liach commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,7 +283,43 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+            return newState;
+        }
+
+        public void setNewState(IBlockState state){
```
brace

liach commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,7 +283,43 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+
+        public IBlockState getNewState()
+        {
+            if (this.isCanceled()) return this.getWorld().getBlockState(this.getPos());
```
this usage does not seem very fine

MoxLotus replied:
Could you elaborate for me?

mezz replied:
You should store the original state in a field and add a getter for it.
For this line, I'm not sure it makes sense to return the original state when the event is canceled. Usually canceled just means "nobody after me gets the event", having it return the original state here is a bit unexpected I think.

MoxLotus replied:
Well, alright.

liach commented:
Can you show a screenshot of the test mod?

MoxLotus commented:
@liach Was there something specific you wanted to see?

liach commented:
just show that obsidians are replaced by gold blocks, etc

akros2 committed:
Bugfix

Missed a line of code where fire was placed by lava. Tweaked the test
mod.

MoxLotus commented:
![2017-12-29_20 58 18](https://user-images.githubusercontent.com/5196430/34451166-62c74052-ecdb-11e7-9ca8-b17b389e263e.png)

akros2 committed:
Canceling the event now works as expected.

Also added more documentation.

MoxLotus commented:
@mezz Have I made all the changes you want?

MoxLotus committed:
Merge branch '1.12.x' into 1.12.x

akros2 committed:
Restoring functionality lost during the last merge.

bs2609 commented:
Needs an update to resolve conflicts.

MoxLotus committed:
Merge branch '1.12.x' into 1.12.x

williewillus commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -282,7 +283,60 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+        }
+
+        /**
+         * @return The original state, as opposed to what minecraft would have placed by default ({@link #getState()}).
```
this is a bit unclear. this *is* the original state that mc would have placed, right? so the words "as opposed to" should not be there.

MoxLotus replied:
It's meant to be the original block, from before the event fired. Will rewrite.

akros2 committed:
Clarified a javadoc comment

LexManos commented:
Thanks. This pull request has been integrated manually.
